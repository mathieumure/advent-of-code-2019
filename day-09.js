import {readFile} from 'fs'
import {join} from 'path'
import {promisify} from 'util'
import {computeIntProgram, computeIntProgramWithState} from "./utils/intProgram";

const asyncReadFile = promisify(readFile)

export const parse = async path => {
    const data = await asyncReadFile(join(__dirname, path));
    const input = data.toString()
        .split('\n')[0]
        .split(',')
        .map(it => BigInt(it));

    return input;
}

export const execIntProgram = (program, input = []) => {
    const output = []
    computeIntProgram(program, input, out => {
        output.push(out)
    })
    return output
}


export const computePart1 = items => {
   return execIntProgram(items, [BigInt(1)])
}

export const computePart2 = items => {
    return execIntProgram(items, [BigInt(2)])
}

export const run = async part => {
    const data = await parse(`./data/day-09.txt`);
    const result = part === 'part1' ? computePart1(data) : computePart2(data)
    return result
}