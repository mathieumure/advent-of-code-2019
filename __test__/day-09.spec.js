import {computePart1, execIntProgram, run} from '../day-09'

describe('day-O9 - part 1', () => {

    it('should get a copy of itself', () => {
        const input = [109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99].map(it => BigInt(it))
        expect(execIntProgram(input).map(it => it.toString())).toEqual(input.map(it => it.toString()))
    })

    it('should output 16 digit number', () => {
        const input = [1102,34915192,34915192,7,4,7,99,0].map(it => BigInt(it))
        expect(execIntProgram(input)[0].toString()).toHaveLength(16)
    })

    it('should output large number', () => {
        const input = [BigInt(104),BigInt(1125899906842624),BigInt(99)]
        expect(execIntProgram(input).toString()).toEqual(BigInt(1125899906842624).toString())
    })

    it('should run', async () => {
        const result = await run('part1')
        expect(result[0]).toMatchSnapshot()
    })
})

describe('day-O9 - part 2', () => {
  it('should run', async () => {
      const result = await run('part2')
      expect(result).toMatchSnapshot()
  })
})
