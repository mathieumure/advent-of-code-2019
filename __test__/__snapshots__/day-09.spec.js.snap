// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`day-O9 - part 1 should run 1`] = `3638931938n`;

exports[`day-O9 - part 2 should run 1`] = `
Array [
  86025n,
]
`;
